// @ls1intum/apollon/lib/components/theme/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var styled_components_1 = tslib_1.__importStar(require("styled-components"));
var styled_components_2 = require("styled-components");
exports.withTheme = styled_components_2.withTheme;
exports.styled = styled_components_1.default;
exports.css = styled_components_1.css;
exports.defaults = {
    color: {
        primary: '#2a8fbd',
        secondary: '#6c757d',
        white: '#fff',
        gray100: '#f8f9fa',
        gray200: '#e9ecef',
        gray300: '#dee2e6',
        gray400: '#ced4da',
        gray500: '#adb5bd',
        gray600: '#6c757d',
        gray700: '#495057',
        gray800: '#343a40',
        gray900: '#212529',
        black: '#212529',
    },
    font: {
        color: '#212529',
        family: 'Helvetica Neue, Helvetica, Arial, sans-serif',
        size: 16,
    },
    interactive: {
        normal: 'rgba(0, 220, 0, 0.3)',
        hovered: 'rgba(0, 220, 0, 0.15)',
    },
};
