// @ls1intum/apollon/lib/packages/common/uml-package/uml-package.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var uml_container_1 = require("../../../services/uml-container/uml-container");
var boundary_1 = require("../../../utils/geometry/boundary");
var text_1 = require("../../../utils/svg/text");
var UMLPackage = /** @class */ (function (_super) {
    tslib_1.__extends(UMLPackage, _super);
    function UMLPackage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UMLPackage.prototype.render = function (layer, children) {
        var _this = this;
        if (children === void 0) { children = []; }
        var radix = 10;
        var nameBounds = {
            x: this.bounds.x,
            y: this.bounds.y,
            width: Math.round((text_1.Text.size(layer, this.name, { fontWeight: 'bold' }).width + 20) / radix) * radix,
            height: 20,
        };
        var absoluteElements = children.map(function (element) {
            element.bounds.x += _this.bounds.x;
            element.bounds.y += _this.bounds.y;
            return element;
        });
        var bounds = boundary_1.computeBoundingBoxForElements(tslib_1.__spread([this, { bounds: nameBounds }], absoluteElements));
        var relativeElements = absoluteElements.map(function (element) {
            element.bounds.x -= _this.bounds.x;
            element.bounds.y -= _this.bounds.y;
            return element;
        });
        var deltaX = bounds.x - this.bounds.x;
        var deltaY = bounds.y - this.bounds.y;
        relativeElements.forEach(function (child) {
            child.bounds.x -= deltaX;
            child.bounds.y -= deltaY;
        });
        this.bounds = bounds;
        return tslib_1.__spread([this], relativeElements);
    };
    return UMLPackage;
}(uml_container_1.UMLContainer));
exports.UMLPackage = UMLPackage;
