// @ls1intum/apollon/lib/components/uml-element/updatable/updatable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_dom_1 = require("react-dom");
var react_redux_1 = require("react-redux");
var uml_element_repository_1 = require("../../../services/uml-element/uml-element-repository");
var initialState = {};
var enhance = react_redux_1.connect(null, {
    updateStart: uml_element_repository_1.UMLElementRepository.updateStart,
});
exports.updatable = function (WrappedComponent) {
    var Updatable = /** @class */ (function (_super) {
        tslib_1.__extends(Updatable, _super);
        function Updatable() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = initialState;
            _this.onDoubleClick = function () {
                _this.props.updateStart(_this.props.id);
            };
            return _this;
        }
        Updatable.prototype.componentDidMount = function () {
            var node = react_dom_1.findDOMNode(this);
            node.addEventListener('dblclick', this.onDoubleClick);
        };
        Updatable.prototype.componentWillUnmount = function () {
            var node = react_dom_1.findDOMNode(this);
            node.removeEventListener('dblclick', this.onDoubleClick);
        };
        Updatable.prototype.render = function () {
            var _a = this.props, updateStart = _a.updateStart, props = tslib_1.__rest(_a, ["updateStart"]);
            return react_1.default.createElement(WrappedComponent, tslib_1.__assign({}, props));
        };
        return Updatable;
    }(react_1.Component));
    return enhance(Updatable);
};
