// @ls1intum/apollon/lib/services/uml-element/uml-element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var uml_element_type_1 = require("../../packages/uml-element-type");
var assign_1 = require("../../utils/fx/assign");
var point_1 = require("../../utils/geometry/point");
var uuid_1 = require("../../utils/uuid");
var uml_element_port_1 = require("./uml-element-port");
/** Class implementation of `IUMLElement` to use inheritance at runtime */
var UMLElement = /** @class */ (function () {
    function UMLElement(values) {
        this.id = uuid_1.uuid();
        this.name = '';
        this.bounds = { x: 0, y: 0, width: 200, height: 100 };
        this.owner = null;
        assign_1.assign(this, values);
    }
    /**
     * Clones an instance of `UMLElement`
     *
     * @param override - Override existing properties.
     */
    UMLElement.prototype.clone = function (override) {
        var Constructor = this.constructor;
        var values = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, this), override), { id: uuid_1.uuid() });
        return new Constructor(values);
    };
    /** Serializes an `UMLElement` to an `Apollon.UMLElement` */
    UMLElement.prototype.serialize = function (children) {
        return {
            id: this.id,
            name: this.name,
            type: this.type,
            owner: this.owner,
            bounds: this.bounds,
            highlight: this.highlight,
        };
    };
    /** Deserializes an `Apollon.UMLElement` to an `UMLElement` */
    UMLElement.prototype.deserialize = function (values, children) {
        this.id = values.id;
        this.name = values.name;
        this.type = values.type;
        this.owner = values.owner || null;
        this.bounds = tslib_1.__assign({}, values.bounds);
        this.highlight = values.highlight;
    };
    /** Calculate the position of each port for the element. */
    UMLElement.prototype.ports = function () {
        var _a;
        return _a = {},
            _a[uml_element_port_1.Direction.Up] = new point_1.Point(this.bounds.width / 2, 0),
            _a[uml_element_port_1.Direction.Right] = new point_1.Point(this.bounds.width, this.bounds.height / 2),
            _a[uml_element_port_1.Direction.Down] = new point_1.Point(this.bounds.width / 2, this.bounds.height),
            _a[uml_element_port_1.Direction.Left] = new point_1.Point(0, this.bounds.height / 2),
            _a;
    };
    /** `UMLElement` type specific feature flags */
    UMLElement.features = {
        connectable: true,
        droppable: false,
        hoverable: true,
        movable: true,
        resizable: true,
        selectable: true,
        updatable: true,
        alternativePortVisualization: false,
    };
    /** Checks whether an `IUMLElement` is of type `UMLElementType` */
    UMLElement.isUMLElement = function (element) {
        return element.type in uml_element_type_1.UMLElementType;
    };
    return UMLElement;
}());
exports.UMLElement = UMLElement;
