// @ls1intum/apollon/lib/components/assessment/assessment-section.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_redux_1 = require("react-redux");
var redux_1 = require("redux");
var assessment_repository_1 = require("../../services/assessment/assessment-repository");
var divider_1 = require("../controls/divider/divider");
var textfield_1 = require("../controls/textfield/textfield");
var typography_1 = require("../controls/typography/typography");
var localized_1 = require("../i18n/localized");
var styles_1 = require("../theme/styles");
var Flex = styles_1.styled.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  display: flex;\n  align-items: baseline;\n  justify-content: space-between;\n"], ["\n  display: flex;\n  align-items: baseline;\n  justify-content: space-between;\n"])));
var enhance = redux_1.compose(localized_1.localized, react_redux_1.connect(function (state, props) { return ({
    readonly: state.editor.readonly,
    assessment: assessment_repository_1.AssessmentRepository.getById(state.assessments)(props.element.id),
    diagramType: state.diagram.type,
}); }, { assess: assessment_repository_1.AssessmentRepository.assess }));
var AssessmentSectionCompoennt = /** @class */ (function (_super) {
    tslib_1.__extends(AssessmentSectionCompoennt, _super);
    function AssessmentSectionCompoennt() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.updateScore = function (value) {
            var _a = _this.props, element = _a.element, assessment = _a.assessment;
            var score = parseFloat(value) || 0;
            _this.props.assess(element.id, tslib_1.__assign(tslib_1.__assign({}, assessment), { score: score }));
        };
        _this.updateFeedback = function (value) {
            var _a = _this.props, element = _a.element, assessment = _a.assessment;
            var feedback = value.length ? value : undefined;
            _this.props.assess(element.id, tslib_1.__assign(tslib_1.__assign({ score: 0 }, assessment), { feedback: feedback }));
        };
        return _this;
    }
    AssessmentSectionCompoennt.prototype.render = function () {
        var _a = this.props, element = _a.element, assessment = _a.assessment, readonly = _a.readonly, diagramType = _a.diagramType;
        return (react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement("section", null,
                react_1.default.createElement(typography_1.Header, null,
                    this.props.translate('assessment.assessment'),
                    ' ',
                    this.props.translate("packages." + diagramType + "." + element.type),
                    element.name ? (react_1.default.createElement(react_1.default.Fragment, null,
                        ' ',
                        react_1.default.createElement("span", { style: { display: 'inline-block' } }, "\"" + element.name + "\""))) : (''))),
            react_1.default.createElement("section", null,
                react_1.default.createElement(Flex, null,
                    react_1.default.createElement("span", { style: { marginRight: '0.5em' } },
                        this.props.translate('assessment.score'),
                        ":"),
                    readonly ? (react_1.default.createElement("span", null, (assessment && assessment.score) || '-')) : (react_1.default.createElement(textfield_1.Textfield, { gutter: true, type: "number", step: 0.5, onChange: this.updateScore, value: assessment ? String(assessment.score) : '' })))),
            readonly ? (assessment && assessment.feedback && react_1.default.createElement("section", null, assessment.feedback)) : (react_1.default.createElement("section", null,
                react_1.default.createElement(textfield_1.Textfield, { multiline: true, placeholder: this.props.translate('assessment.feedback'), onChange: this.updateFeedback, enterToSubmit: false, value: assessment && assessment.feedback ? assessment.feedback : '' }))),
            react_1.default.createElement(divider_1.Divider, null)));
    };
    return AssessmentSectionCompoennt;
}(react_1.Component));
exports.AssessmentSection = enhance(AssessmentSectionCompoennt);
var templateObject_1;
