// @ls1intum/apollon/lib/utils/actions/sagas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var symbols_1 = require("@redux-saga/symbols");
var effects_1 = require("redux-saga/effects");
exports.isInternal = function (action) {
    return symbols_1.SAGA_ACTION in action;
};
function composeSaga(sagas) {
    return effects_1.all(sagas.map(effects_1.spawn));
}
exports.composeSaga = composeSaga;
function run(sagas) {
    return effects_1.all(sagas.map(function (saga) { return exports.keepAlive(exports.safely(saga)); }));
}
exports.run = run;
exports.keepAlive = function (saga) {
    return effects_1.spawn(function () {
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!true) return [3 /*break*/, 2];
                    return [4 /*yield*/, effects_1.call(saga)];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 0];
                case 2: return [2 /*return*/];
            }
        });
    });
};
exports.safely = function (saga) {
    function safelySaga() {
        var e_1;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, effects_1.call(saga)];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    e_1 = _a.sent();
                    // tslint:disable-next-line
                    console.error(e_1);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }
    return safelySaga;
};
