// @ls1intum/apollon/lib/components/uml-element/assessable/assessment-styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var check_1 = require("../../controls/icon/check");
var exclamation_1 = require("../../controls/icon/exclamation");
var times_1 = require("../../controls/icon/times");
var styles_1 = require("../../theme/styles");
var ICON_SIZE = 24;
exports.Container = styles_1.styled.circle.attrs(function (props) { return ({
    r: ICON_SIZE / 2 + 4,
    fillOpacity: 0.8,
    fill: props.theme.color.gray200,
}); })(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject([""], [""])));
var icon = {
    x: -ICON_SIZE / 2,
    y: -ICON_SIZE / 2,
    width: ICON_SIZE,
    height: ICON_SIZE,
};
exports.CorrectIcon = styles_1.styled(check_1.CheckIcon).attrs(icon)(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n  fill: green;\n"], ["\n  fill: green;\n"])));
exports.WrongIcon = styles_1.styled(times_1.TimesIcon).attrs(icon)(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n  fill: red;\n"], ["\n  fill: red;\n"])));
exports.FeedbackIcon = styles_1.styled(exclamation_1.ExclamationIcon).attrs(icon)(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n  fill: blue;\n"], ["\n  fill: blue;\n"])));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
