// @ls1intum/apollon/lib/components/uml-element/assessable/assessable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_redux_1 = require("react-redux");
var uml_relationship_1 = require("../../../services/uml-relationship/uml-relationship");
var path_1 = require("../../../utils/geometry/path");
var point_1 = require("../../../utils/geometry/point");
var assessment_styles_1 = require("./assessment-styles");
var react_dom_1 = require("react-dom");
var uml_element_repository_1 = require("../../../services/uml-element/uml-element-repository");
var assessment_repository_1 = require("../../../services/assessment/assessment-repository");
var enhance = react_redux_1.connect(function (state, props) {
    var element = state.elements[props.id];
    return {
        assessment: state.assessments[props.id],
        bounds: element.bounds,
        path: uml_relationship_1.UMLRelationship.isUMLRelationship(element) ? element.path : undefined,
        readonly: state.editor.readonly,
    };
}, {
    select: uml_element_repository_1.UMLElementRepository.select,
    deselect: uml_element_repository_1.UMLElementRepository.deselect,
    assess: assessment_repository_1.AssessmentRepository.assess,
    updateStart: uml_element_repository_1.UMLElementRepository.updateStart,
});
exports.assessable = function (WrappedComponent) {
    var Assessable = /** @class */ (function (_super) {
        tslib_1.__extends(Assessable, _super);
        function Assessable() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.onDragOver = function (ev) {
                // prevent default to allow drop
                ev.preventDefault();
                // don't propagate to parents, so that most accurate element is selected only
                ev.stopPropagation();
                _this.props.select(_this.props.id);
            };
            _this.onDragLeave = function () {
                _this.props.deselect(_this.props.id);
            };
            /**
             * Artemis instruction object can be dropped on assessment sections to automatically fill assessment
             * @param ev DropEvent
             */
            _this.onDrop = function (ev) {
                // prevent default action (open as link for some elements)
                ev.preventDefault();
                // unselect current element
                _this.props.deselect(_this.props.id);
                // don't propagate to parents, so that most accurate element is selected only
                ev.stopPropagation();
                if (!!ev.dataTransfer) {
                    var data = ev.dataTransfer.getData('artemis/sgi');
                    if (!data) {
                        console.warn('Could not get artemis sgi element from drop element');
                        return;
                    }
                    var instruction = void 0;
                    try {
                        instruction = JSON.parse(data);
                    }
                    catch (e) {
                        console.error('Could not parse artemis sgi', e);
                        return;
                    }
                    var _a = _this.props, elementId = _a.id, assessment = _a.assessment;
                    var score = instruction.credits;
                    var feedback = instruction.feedback;
                    _this.props.assess(elementId, tslib_1.__assign(tslib_1.__assign({}, assessment), { score: score, feedback: feedback }));
                    _this.props.updateStart(elementId);
                }
            };
            return _this;
        }
        Assessable.prototype.componentDidMount = function () {
            if (!this.props.readonly) {
                var node = react_dom_1.findDOMNode(this);
                node.addEventListener('dragover', this.onDragOver.bind(this));
                node.addEventListener('dragleave', this.onDragLeave.bind(this));
                node.addEventListener('drop', this.onDrop.bind(this));
            }
        };
        Assessable.prototype.componentWillUnmount = function () {
            var node = react_dom_1.findDOMNode(this);
            node.removeEventListener('dragover', this.onDragOver);
            node.removeEventListener('dragleave', this.onDragLeave);
            node.removeEventListener('drop', this.onDrop);
        };
        Assessable.prototype.render = function () {
            var _a = this.props, assessment = _a.assessment, assess = _a.assess, select = _a.select, deselect = _a.deselect, updateStart = _a.updateStart, bounds = _a.bounds, ipath = _a.path, readonly = _a.readonly, props = tslib_1.__rest(_a, ["assessment", "assess", "select", "deselect", "updateStart", "bounds", "path", "readonly"]);
            var position;
            if (ipath) {
                var path = new path_1.Path(ipath);
                position = path.position(path.length / 2);
            }
            else {
                position = new point_1.Point(bounds.width, 0);
            }
            return (react_1.default.createElement(WrappedComponent, tslib_1.__assign({}, props), assessment && (react_1.default.createElement("g", { transform: "translate(" + position.x + " " + position.y + ")", pointerEvents: 'none' },
                assessment.score === 0 && !!assessment.feedback && (react_1.default.createElement(react_1.default.Fragment, null,
                    react_1.default.createElement(assessment_styles_1.Container, null),
                    react_1.default.createElement(assessment_styles_1.FeedbackIcon, null))),
                assessment.score > 0 && (react_1.default.createElement(react_1.default.Fragment, null,
                    react_1.default.createElement(assessment_styles_1.Container, null),
                    react_1.default.createElement(assessment_styles_1.CorrectIcon, null))),
                assessment.score < 0 && (react_1.default.createElement(react_1.default.Fragment, null,
                    react_1.default.createElement(assessment_styles_1.Container, null),
                    react_1.default.createElement(assessment_styles_1.WrongIcon, null)))))));
        };
        return Assessable;
    }(react_1.Component));
    return enhance(Assessable);
};
