// @ls1intum/apollon/lib/components/connectable/uml-relationship-preview.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_redux_1 = require("react-redux");
var uml_element_repository_1 = require("../../services/uml-element/uml-element-repository");
var point_1 = require("../../utils/geometry/point");
var enhance = react_redux_1.connect(function (state, props) { return ({
    ports: uml_element_repository_1.UMLElementRepository.get(state.elements[props.port.element]).ports(),
}); }, {
    end: uml_element_repository_1.UMLElementRepository.endConnecting,
    getAbsolutePosition: uml_element_repository_1.UMLElementRepository.getAbsolutePosition,
});
var RelationshipPreview = /** @class */ (function (_super) {
    tslib_1.__extends(RelationshipPreview, _super);
    function RelationshipPreview() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RelationshipPreview.prototype.render = function () {
        var _a = this.props, port = _a.port, ports = _a.ports;
        var _b = this.props.getAbsolutePosition(port.element), x = _b.x, y = _b.y;
        var position = tslib_1.__assign({}, ports[port.direction]);
        var source = new point_1.Point(x + position.x, y + position.y);
        var path = [source, this.props.target];
        var points = path.map(function (p) { return p.x + " " + p.y; }).join(', ');
        return (react_1.default.createElement("polyline", { points: points, pointerEvents: "none", fill: "none", stroke: "black", strokeWidth: "1", strokeDasharray: "5,5" }));
    };
    return RelationshipPreview;
}(react_1.Component));
exports.UMLRelationshipPreview = enhance(RelationshipPreview);
