// @ls1intum/apollon/lib/components/uml-element/movable/movable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_dom_1 = require("react-dom");
var react_redux_1 = require("react-redux");
var uml_element_repository_1 = require("../../../services/uml-element/uml-element-repository");
var point_1 = require("../../../utils/geometry/point");
var initialState = {
    offset: new point_1.Point(),
};
var enhance = react_redux_1.connect(function (state, props) { return ({
    movable: state.selected.includes(props.id) && !state.resizing.includes(props.id) && !state.connecting.length,
    moving: state.moving.includes(props.id),
}); }, {
    start: uml_element_repository_1.UMLElementRepository.startMoving,
    move: uml_element_repository_1.UMLElementRepository.move,
    end: uml_element_repository_1.UMLElementRepository.endMoving,
});
exports.movable = function (WrappedComponent) {
    var Movable = /** @class */ (function (_super) {
        tslib_1.__extends(Movable, _super);
        function Movable() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = initialState;
            _this.move = function (x, y) {
                x = Math.round(x / 10) * 10;
                y = Math.round(y / 10) * 10;
                if (x === 0 && y === 0)
                    return;
                _this.setState(function (state) { return ({ offset: state.offset.add(x, y) }); });
                _this.props.move({ x: x, y: y });
            };
            _this.onPointerDown = function (event) {
                if (event.which && event.which !== 1) {
                    return;
                }
                _this.setState({ offset: new point_1.Point(event.clientX, event.clientY) });
                document.addEventListener('pointermove', _this.onPointerMove);
                document.addEventListener('pointerup', _this.onPointerUp, { once: true });
                setTimeout(function () { return !_this.props.movable && _this.onPointerUp(); }, 0);
            };
            _this.onPointerMove = function (event) {
                var x = event.clientX - _this.state.offset.x;
                var y = event.clientY - _this.state.offset.y;
                if (!_this.props.moving) {
                    if (Math.abs(x) > 5 || Math.abs(y) > 5) {
                        _this.props.start();
                    }
                }
                else {
                    _this.move(x, y);
                }
            };
            _this.onPointerUp = function () {
                document.removeEventListener('pointermove', _this.onPointerMove);
                if (!_this.props.moving) {
                    return;
                }
                _this.setState(initialState);
                _this.props.end();
            };
            return _this;
        }
        Movable.prototype.componentDidMount = function () {
            var node = react_dom_1.findDOMNode(this);
            node.style.cursor = 'move';
            var child = node.firstChild;
            child.addEventListener('pointerdown', this.onPointerDown);
        };
        Movable.prototype.componentWillUnmount = function () {
            var node = react_dom_1.findDOMNode(this);
            var child = node.firstChild;
            child.removeEventListener('pointerdown', this.onPointerDown);
            document.removeEventListener('pointermove', this.onPointerMove);
            document.removeEventListener('pointerup', this.onPointerUp);
        };
        Movable.prototype.render = function () {
            var _a = this.props, _ = _a.movable, start = _a.start, move = _a.move, end = _a.end, props = tslib_1.__rest(_a, ["movable", "start", "move", "end"]);
            return react_1.default.createElement(WrappedComponent, tslib_1.__assign({}, props));
        };
        return Movable;
    }(react_1.Component));
    return enhance(Movable);
};
