// @ls1intum/apollon/lib/services/uml-element/resizable/resizable-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Resizable = {
    startResizing: function (id) { return function (dispatch, getState) {
        var ids = id ? (Array.isArray(id) ? id : [id]) : getState().selected;
        if (!ids.length) {
            return;
        }
        dispatch({
            type: "@@element/resizable/START" /* START */,
            payload: { ids: ids },
            undoable: true,
        });
    }; },
    resize: function (delta, id) { return function (dispatch, getState) {
        var ids = id ? (Array.isArray(id) ? id : [id]) : getState().resizing;
        if (!ids.length) {
            return;
        }
        dispatch({
            type: "@@element/resizable/RESIZE" /* RESIZE */,
            payload: { ids: ids, delta: delta },
            undoable: false,
        });
    }; },
    endResizing: function (id) { return function (dispatch, getState) {
        var ids = id ? (Array.isArray(id) ? id : [id]) : getState().resizing;
        if (!ids.length) {
            return;
        }
        dispatch({
            type: "@@element/resizable/END" /* END */,
            payload: { ids: ids },
            undoable: false,
        });
    }; },
};
