// @ls1intum/apollon/lib/components/uml-element/hoverable/hoverable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_dom_1 = require("react-dom");
var react_redux_1 = require("react-redux");
var uml_element_repository_1 = require("../../../services/uml-element/uml-element-repository");
var enhance = react_redux_1.connect(null, {
    hover: uml_element_repository_1.UMLElementRepository.hover,
    leave: uml_element_repository_1.UMLElementRepository.leave,
});
exports.hoverable = function (WrappedComponent) {
    var Hoverable = /** @class */ (function (_super) {
        tslib_1.__extends(Hoverable, _super);
        function Hoverable() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.enter = function () { return _this.props.hover(_this.props.id); };
            _this.leave = function () { return _this.props.leave(_this.props.id); };
            return _this;
        }
        Hoverable.prototype.componentDidMount = function () {
            var node = react_dom_1.findDOMNode(this);
            node.addEventListener('pointerenter', this.enter);
            node.addEventListener('pointerleave', this.leave);
        };
        Hoverable.prototype.componentWillUnmount = function () {
            var node = react_dom_1.findDOMNode(this);
            node.removeEventListener('pointerenter', this.enter);
            node.removeEventListener('pointerleave', this.leave);
        };
        Hoverable.prototype.render = function () {
            var _a = this.props, hover = _a.hover, leave = _a.leave, props = tslib_1.__rest(_a, ["hover", "leave"]);
            return react_1.default.createElement(WrappedComponent, tslib_1.__assign({}, props));
        };
        return Hoverable;
    }(react_1.Component));
    return enhance(Hoverable);
};
