// @ls1intum/apollon/lib/packages/uml-use-case-diagram/uml-use-case-association/uml-use-case-association-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_redux_1 = require("react-redux");
var redux_1 = require("redux");
var styled_components_1 = tslib_1.__importDefault(require("styled-components"));
var __1 = require("..");
var button_1 = require("../../../components/controls/button/button");
var divider_1 = require("../../../components/controls/divider/divider");
var dropdown_1 = require("../../../components/controls/dropdown/dropdown");
var exchange_1 = require("../../../components/controls/icon/exchange");
var trash_1 = require("../../../components/controls/icon/trash");
var textfield_1 = require("../../../components/controls/textfield/textfield");
var typography_1 = require("../../../components/controls/typography/typography");
var localized_1 = require("../../../components/i18n/localized");
var uml_element_repository_1 = require("../../../services/uml-element/uml-element-repository");
var uml_relationship_repository_1 = require("../../../services/uml-relationship/uml-relationship-repository");
var Flex = styled_components_1.default.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  display: flex;\n  align-items: baseline;\n  justify-content: space-between;\n"], ["\n  display: flex;\n  align-items: baseline;\n  justify-content: space-between;\n"])));
var UseCaseAssociationUpdate = /** @class */ (function (_super) {
    tslib_1.__extends(UseCaseAssociationUpdate, _super);
    function UseCaseAssociationUpdate() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.rename = function (id) { return function (value) {
            _this.props.update(id, { name: value });
        }; };
        _this.onChange = function (type) {
            var _a = _this.props, element = _a.element, update = _a.update;
            update(element.id, { type: type });
        };
        return _this;
    }
    UseCaseAssociationUpdate.prototype.render = function () {
        var _a;
        var _this = this;
        var element = this.props.element;
        return (react_1.default.createElement("div", null,
            react_1.default.createElement("section", null, element.type === __1.UseCaseRelationshipType.UseCaseAssociation ? (react_1.default.createElement(Flex, null,
                react_1.default.createElement(textfield_1.Textfield, { value: element.name, placeholder: "...", onChange: this.rename(element.id), autoFocus: true }),
                react_1.default.createElement(button_1.Button, { color: "link", tabIndex: -1, onClick: function () { return _this.props.delete(element.id); } },
                    react_1.default.createElement(trash_1.TrashIcon, null)))) : (react_1.default.createElement(Flex, null,
                react_1.default.createElement(typography_1.Header, { gutter: false, style: { flexGrow: 1 } }, (_a = {},
                    _a[__1.UseCaseRelationshipType.UseCaseAssociation] = this.props.translate('packages.UseCaseDiagram.UseCaseAssociation'),
                    _a[__1.UseCaseRelationshipType.UseCaseGeneralization] = this.props.translate('packages.UseCaseDiagram.UseCaseGeneralization'),
                    _a[__1.UseCaseRelationshipType.UseCaseInclude] = this.props.translate('packages.UseCaseDiagram.UseCaseInclude'),
                    _a[__1.UseCaseRelationshipType.UseCaseExtend] = this.props.translate('packages.UseCaseDiagram.UseCaseExtend'),
                    _a)[element.type]),
                react_1.default.createElement(button_1.Button, { color: "link", tabIndex: -1, onClick: function () { return _this.props.flip(element.id); } },
                    react_1.default.createElement(exchange_1.ExchangeIcon, null)),
                react_1.default.createElement(button_1.Button, { color: "link", tabIndex: -1, onClick: function () { return _this.props.delete(element.id); } },
                    react_1.default.createElement(trash_1.TrashIcon, null))))),
            react_1.default.createElement("section", null,
                react_1.default.createElement(divider_1.Divider, null),
                react_1.default.createElement(dropdown_1.Dropdown, { value: element.type, onChange: this.onChange },
                    react_1.default.createElement(dropdown_1.Dropdown.Item, { value: __1.UseCaseRelationshipType.UseCaseAssociation }, this.props.translate('packages.UseCaseDiagram.UseCaseAssociation')),
                    react_1.default.createElement(dropdown_1.Dropdown.Item, { value: __1.UseCaseRelationshipType.UseCaseGeneralization }, this.props.translate('packages.UseCaseDiagram.UseCaseGeneralization')),
                    react_1.default.createElement(dropdown_1.Dropdown.Item, { value: __1.UseCaseRelationshipType.UseCaseInclude }, this.props.translate('packages.UseCaseDiagram.UseCaseInclude')),
                    react_1.default.createElement(dropdown_1.Dropdown.Item, { value: __1.UseCaseRelationshipType.UseCaseExtend }, this.props.translate('packages.UseCaseDiagram.UseCaseExtend'))))));
    };
    return UseCaseAssociationUpdate;
}(react_1.Component));
var enhance = redux_1.compose(localized_1.localized, react_redux_1.connect(null, {
    update: uml_element_repository_1.UMLElementRepository.update,
    delete: uml_element_repository_1.UMLElementRepository.delete,
    flip: uml_relationship_repository_1.UMLRelationshipRepository.flip,
}));
exports.UMLUseCaseAssociationUpdate = enhance(UseCaseAssociationUpdate);
var templateObject_1;
