// @ls1intum/apollon/lib/components/create-pane/create-pane.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_redux_1 = require("react-redux");
var redux_1 = require("redux");
var diagram_type_1 = require("../../packages/diagram-type");
var activity_preview_1 = require("../../packages/uml-activity-diagram/activity-preview");
var class_preview_1 = require("../../packages/uml-class-diagram/class-preview");
var communication_preview_1 = require("../../packages/uml-communication-diagram/communication-preview");
var component_preview_1 = require("../../packages/uml-component-diagram/component-preview");
var deployment_preview_1 = require("../../packages/uml-deployment-diagram/deployment-preview");
var object_preview_1 = require("../../packages/uml-object-diagram/object-preview");
var use_case_preview_1 = require("../../packages/uml-use-case-diagram/use-case-preview");
var uml_element_repository_1 = require("../../services/uml-element/uml-element-repository");
var tree_1 = require("../../utils/geometry/tree");
var with_canvas_1 = require("../canvas/with-canvas");
var localized_1 = require("../i18n/localized");
var model_store_1 = require("../store/model-store");
var preview_element_1 = require("./preview-element");
var getInitialState = function (_a) {
    var type = _a.type, canvas = _a.canvas, translate = _a.translate;
    var previews = [];
    switch (type) {
        case diagram_type_1.UMLDiagramType.ClassDiagram:
            previews.push.apply(previews, tslib_1.__spread(class_preview_1.composeClassPreview(canvas, translate)));
            break;
        case diagram_type_1.UMLDiagramType.ObjectDiagram:
            previews.push.apply(previews, tslib_1.__spread(object_preview_1.composeObjectPreview(canvas, translate)));
            break;
        case diagram_type_1.UMLDiagramType.ActivityDiagram:
            previews.push.apply(previews, tslib_1.__spread(activity_preview_1.composeActivityPreview(canvas, translate)));
            break;
        case diagram_type_1.UMLDiagramType.UseCaseDiagram:
            previews.push.apply(previews, tslib_1.__spread(use_case_preview_1.composeUseCasePreview(canvas, translate)));
            break;
        case diagram_type_1.UMLDiagramType.CommunicationDiagram:
            previews.push.apply(previews, tslib_1.__spread(communication_preview_1.composeCommunicationPreview(canvas, translate)));
            break;
        case diagram_type_1.UMLDiagramType.ComponentDiagram:
            previews.push.apply(previews, tslib_1.__spread(component_preview_1.composeComponentPreview(canvas, translate)));
            break;
        case diagram_type_1.UMLDiagramType.DeploymentDiagram:
            previews.push.apply(previews, tslib_1.__spread(deployment_preview_1.composeDeploymentPreview(canvas, translate)));
            break;
    }
    return { previews: previews };
};
var enhance = redux_1.compose(localized_1.localized, with_canvas_1.withCanvas, react_redux_1.connect(function (state) { return ({
    type: state.diagram.type,
}); }, {
    create: uml_element_repository_1.UMLElementRepository.create,
}));
var CreatePaneComponent = /** @class */ (function (_super) {
    tslib_1.__extends(CreatePaneComponent, _super);
    function CreatePaneComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = getInitialState(_this.props);
        _this.create = function (preview, owner) {
            var elements = tree_1.clone(preview, _this.state.previews);
            _this.props.create(elements, owner);
        };
        return _this;
    }
    CreatePaneComponent.prototype.render = function () {
        var _this = this;
        var features = {
            hoverable: false,
            selectable: false,
            movable: false,
            resizable: false,
            connectable: false,
            updatable: false,
            droppable: false,
            alternativePortVisualization: false,
        };
        var previews = this.state.previews;
        var elements = previews.reduce(function (state, preview) {
            var _a;
            return (tslib_1.__assign(tslib_1.__assign({}, state), (_a = {}, _a[preview.id] = tslib_1.__assign({}, preview), _a)));
        }, {});
        return (react_1.default.createElement(model_store_1.StoreProvider, { initialState: { elements: elements, editor: { features: features } } }, Object.values(previews)
            .filter(function (preview) { return !preview.owner; })
            .map(function (preview, index) { return (react_1.default.createElement(preview_element_1.PreviewElement, { key: index, element: preview, create: _this.create })); })));
    };
    return CreatePaneComponent;
}(react_1.Component));
exports.CreatePane = enhance(CreatePaneComponent);
